login

```diff
@@ -2,62 +2,62 @@
 
 # API documentation: http://www.mediawiki.org/wiki/API:Main_page
 
-stty -echo
-read -p "Password: " USERPASS
-stty echo
-echo
+source wiki-config
 
 if [ -z "$USERPASS" ]; then
-  echo "Password not supplied"
+  echo "ERROR: Password not supplied!"
   exit 1
 fi
 
-. wiki-config
- 
-echo "Logging into $WIKIAPI as $USERNAME..."
-
-CR=$(curl -s -S \
-	--location \
-	--retry 2 \
-	--retry-delay 5\
-	--cookie $cookie_jar \
-	--cookie-jar $cookie_jar \
-	--keepalive-time 60 \
-	--header "Accept-Language: en-GB" \
-	--header "Connection: keep-alive" \
-	--compressed \
-	--data-urlencode "lgname=${USERNAME}" \
-	--data-urlencode "lgpassword=${USERPASS}" \
-	--data-urlencode "lgdomain=${USERDOMAIN}" \
-	--request "POST" "${WIKIAPI}?action=login&format=txt")
-
-
-TOKEN=$(echo "$CR" | awk '/\[token\] =>/ {print $3}')
-RESULT=$(echo "$CR" | awk '/\[result\] =>/ {print $3}')
+echo "INFO: Logging into $WIKIAPI as $USERNAME..."
+
+curl \
+   --silent \
+   --show-error \
+   --location \
+   --retry 2 \
+   --retry-delay 5\
+   --cookie "$cookie_jar" \
+   --cookie-jar "$cookie_jar" \
+   --keepalive-time 60 \
+   --header "Accept-Language: en-GB" \
+   --header "Connection: keep-alive" \
+   --compressed \
+   --data-urlencode "lgname=${USERNAME}" \
+   --data-urlencode "lgpassword=${USERPASS}" \
+   --data-urlencode "lgdomain=${USERDOMAIN}" \
+   --output "${TMPFOLDER}/login.json" \
+   --request "POST" \
+   "${WIKIAPI}?action=login&format=json"
+
+RESULT=$(cat "${TMPFOLDER}/login.json" | jq --raw-output '.login.result')
+TOKEN=$(cat "${TMPFOLDER}/login.json" | jq --raw-output '.login.token')
 
 if [ "$RESULT" = "NeedToken" ]; then
-  CR=$(curl -s -S \
-	--location \
-	--cookie $cookie_jar \
-	--cookie-jar $cookie_jar \
-	--keepalive-time 60 \
-	--header "Accept-Language: en-GB" \
-	--header "Connection: keep-alive" \
-	--compressed \
-	--data-urlencode "lgname=${USERNAME}" \
-	--data-urlencode "lgpassword=${USERPASS}" \
-	--data-urlencode "lgdomain=${USERDOMAIN}" \
-	--data-urlencode "lgtoken=${TOKEN}" \
-	--request "POST" "${WIKIAPI}?action=login&format=txt")
-
-  RESULT=$(echo "$CR" | awk '/\[result\] =>/ {print $3}')
+  curl \
+   --silent \
+   --show-error \
+   --location \
+   --cookie "$cookie_jar" \
+   --cookie-jar "$cookie_jar" \
+   --keepalive-time 60 \
+   --header "Accept-Language: en-GB" \
+   --header "Connection: keep-alive" \
+   --compressed \
+   --data-urlencode "lgname=${USERNAME}" \
+   --data-urlencode "lgpassword=${USERPASS}" \
+   --data-urlencode "lgdomain=${USERDOMAIN}" \
+   --data-urlencode "lgtoken=${TOKEN}" \
+   --output "${TMPFOLDER}/login.json2" \
+   --request "POST" \
+   "${WIKIAPI}?action=login&format=json"
+
+   RESULT=$(cat "${TMPFOLDER}/login.json2" | jq -r '.login.result')
 fi
 
 if [ "$RESULT" = "Success" ]; then
-  echo "Successfully logged in as $USERNAME."
+  echo "INFO: Successfully logged in as $USERNAME."
 else
-  echo "Login failed:"
-  echo "$CR"
+  echo "ERROR: Login failed"
```

Comment:
`error:`

Reply:
probably should also be `1>&2` on any errors (or just add an `error()` function to make it neater)

Reply:
Yes.